src/extension/infer.rs

```diff
@@ -1092,4 +1104,183 @@ mod test {
         }
         Ok(())
     }
+
+    fn make_opaque(extension: impl Into<ExtensionId>, signature: FunctionType) -> ops::LeafOp {
+        let opaque =
+            ops::custom::OpaqueOp::new(extension.into(), "", "".into(), vec![], Some(signature));
+        ops::custom::ExternalOp::from(opaque).into()
+    }
+
+    /// A CFG rooted hugr adding resources at each basic block.
+    /// Looks like this:
+    ///
+    ///          +-------------+
+    ///          |    Entry    |
+    ///          |  (Adds [A]) |
+    ///          +-/---------\-+
+    ///           /           \
+    ///  +-------/-----+     +-\-------------+
+    ///  |     BB0     |     |      BB1      |
+    ///  | (Adds [BC]) |     |   (Adds [B])  |
+    ///  +----\--------+     +---/------\----+
+    ///        \                /        \
+    ///         \              /          \
+    ///          \       +----/-------+  +-\---------+
+    ///           \      |   BB10     |  |  BB11     |
+    ///            \     | (Adds [C]) |  | (Adds [C])|
+    ///             \    +----+-------+  +/----------+
+    ///              \        |          /
+    ///         +-----\-------+---------/-+
+    ///         |           Exit          |
+    ///         +-------------------------+
+    #[test]
+    fn infer_cfg_test() -> Result<(), Box<dyn Error>> {
+        let a = ExtensionSet::singleton(&A);
+        let abc = ExtensionSet::from_iter([A, B, C]);
+        let bc = ExtensionSet::from_iter([B, C]);
+        let b = ExtensionSet::singleton(&B);
+        let c = ExtensionSet::singleton(&C);
+
+        let oneway = vec![Type::new_predicate([type_row![NAT]])];
+        let twoway = vec![Type::new_predicate([type_row![NAT], type_row![NAT]])];
+
+        let mut hugr = Hugr::new(NodeType::open_extensions(ops::CFG {
+            inputs: type_row![NAT],
+            outputs: type_row![NAT],
+            extension_delta: abc,
+        }));
+
+        let root = hugr.root();
+
+        let [entry, entry_in, entry_out] = create_with_io(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: a.clone(),
+            },
+        )?;
+
+        let exit = hugr.add_node_with_parent(
+            hugr.root(),
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let mkpred = hugr.add_node_with_parent(
+            entry,
+            NodeType::open_extensions(make_opaque(
+                A,
+                FunctionType::new(vec![NAT], twoway.clone()).with_extension_delta(&a),
+            )),
+        )?;
+
+        let [bb0, bb0_in, bb0_out] = create_with_io(
```

Comment:
You've got four lots here of `create_with_io` + `add_node_with_parent` + 2*`connect` ("internal wiring"). I'd say that was worth making a helper function or lambda, e.g. inside this test. (It can probably create the BasicBlock::DFB as well.)